clients/cobol-lsp-vscode-extension/src/dialect/DialectService.ts

```diff
@@ -0,0 +1,363 @@
+  ) {}
+
+  async resolveCopybook(
+    copybookName: string,
+    statementRange: vscode.Range,
+    nameRange: vscode.Range,
+  ): Promise<
+    | {
+        context: IDocumentProcessingContext;
+        uri: string;
+        text: string;
+      }
+    | undefined
+  > {
+    const statementLocation = new vscode.Location(
+      vscode.Uri.parse(this.documentUri),
+      statementRange,
+    );
+    const nameLocation = new vscode.Location(
+      vscode.Uri.parse(this.documentUri),
+      nameRange,
+    );
+
+    const copybookInfo = await this.dialectService.resolveCopybook(
+      this.programUri,
+      this.dialectName,
+      copybookName,
+      statementLocation,
+      nameLocation,
+    );
+
+    if (copybookInfo) {
+      const copybookContext = new Context(
+        this.dialectService,
+        this.dialectName,
+        this.programUri,
+        copybookInfo.uri,
+        copybookInfo,
+      );
+      this.children.push(copybookContext);
+      copybookContext.copybookInfo = copybookInfo;
+
+      return {
+        context: copybookContext,
+        uri: copybookInfo.uri,
+        text: copybookInfo.text,
+      };
+    }
+    return undefined;
+  }
+  replace(range: vscode.Range, text: string): void {
+    const location = new Location(vscode.Uri.parse(this.documentUri), range);
+    const replacement: DocumentReplacement = {
+      range: location.range,
+      text: text,
+    };
+    this.replacements.push(replacement);
+  }
+}
+
+export class DialectService {
+  private handlers: Map<string, V2StartProcessingHandler> = new Map();
+
+  private diagnosticService: DialectDiagnosticService =
+    new DialectDiagnosticService();
+
+  public constructor(
+    context: vscode.ExtensionContext,
+    languageClientService: LanguageClientService,
+    private copybookService: CopybookDownloadService,
+    private outputChannel?: vscode.OutputChannel,
+  ) {
+    const disposable = vscode.workspace.onDidChangeConfiguration((event) => {
+      if (event.affectsConfiguration(SETTINGS_DIALECT))
+        this.diagnosticService.clear();
+    });
+
+    context.subscriptions.push(disposable);
+
+    languageClientService.addRequestHandler(
+      "dialect/process",
+      async (dialectName: string, programUri: string, text: string) => {
+        const handler = this.handlers.get(dialectName);
+        if (handler) {
+          const context: Context = new Context(
+            this,
+            dialectName,
+            programUri,
+            programUri,
+          );
+
+          try {
+            const diagnostics = await handler(context, programUri, text);
+            this.diagnosticService.publish(
+              dialectName,
+              programUri,
+              diagnostics,
+            );
+          } catch (e) {
+            this.outputChannel?.appendLine(
+              `Dialect ${dialectName} processing fails. Cause: ${JSON.stringify(e)}`,
+            );
+            return undefined;
```

Comment:
After re-reading some of the server side code, maybe this should be a re-throw?
```suggestion
            throw e;
```